pensieve/config.py

```diff
@@ -0,0 +1,160 @@
+import enum
+from typing import Dict, List, Optional
+
+import attr
+import cattr
+import mozanalysis.metrics
+import mozanalysis.metrics.desktop
+import mozanalysis.segments
+
+import pensieve.experimenter
+
+"""
+Spec objects are direct representations of the configuration and contain unresolved references
+to metrics and data sources.
+
+Calling .resolve(config_spec) on a Spec object produces a concrete resolved Conf class.
```

Comment:
I like this concept though it doesn't seem to be applied across the board here.

* `MetricReference` resolves to a `Metric`
* `MetricDefinition` resolves to a `Metric`
* `MetricsSpec` resolves to a `MetricsConfiguration`
* `DataSourceReference` resolves to a `DataSource`
* `DataSourceDefinition` resolves to a `DataSource`
* `ConfigurationSpec` resolves to an `AnalysisConfiguration`. I wonder if `ConfigurationSpec` should be renamed to `AnalysisSpec` since it resolves to `AnalysisConfiguration`, to follow the naming conventions seen above.
-------
* `ExperimentSpec` does not resolve at all? Maybe the `sync()` should actually be a `resolve()` and it resolves to `Experiment`?
* `DataSourcesSpec` does not resolve either? Maybe we don't need a `DataSourcesSpec` or `DataSourcesConfiguration`? It seems that by the time we have an `AnalysisConfiguration` which is what I assume is the final product, we only have an `Experiment` and a `MetricsConfiguration` in it. So data sources are only needed as input to create a `Metric`.

Reply:
Thanks for the careful review! This is a good point; "Reference" and "Definition" objects are also direct representations of the configuration but, when resolved, produce concrete mozanalysis objects instead of Conf objects.

> DataSourcesSpec does not resolve either?

It will need to; I just didn't write it. 😅

Reply:
>  I wonder if ConfigurationSpec should be renamed to AnalysisSpec since it resolves to AnalysisConfiguration

I like this; the current naming is clunky.

Reply:
It might make sense to have something like abstract `Spec` and `Reference` classes. Since, for example, all `...Reference` classes have a `name` attribute and implement the same methods. Might make things a little clearer, however it looks like Pythons support for those things is not that great

Reply:
I thought about implementing a `Resolver` interface with an abstract base class but it seemed like the only purpose would be documentation.

Reply:
>> DataSourcesSpec does not resolve either?

> It will need to; I just didn't write it. 😅

Update: I lied; we don't ever need a concrete DataSourcesConfiguration. Eh! I added an apologetic comment.